codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpProtocolGeneratorUtils.java

```diff
@@ -131,4 +136,54 @@ public static boolean isShapeWithResponseBindings(Model model, Shape shape, Http
         }
         return false;
     }
+
+    /**
+     * Sets the HostPrefix on the request if the operation has the Endpoint trait.
+     *
+     * <p>If there are no HostLabels then this will be a simple string assignment,
+     * otherwise a string builder will be used.
+     *
+     * <p>This assumes that the smithyhttp.Request is available under the variable
+     * "request" and the operation's input struct is available under the variable
+     * "input".
+     *
+     * @param context The generation context.
+     * @param operation The operation to set the host prefix for.
+     */
+    public static void setHostPrefix(GenerationContext context, OperationShape operation) {
+        if (!operation.hasTrait(EndpointTrait.ID)) {
+            return;
+        }
+        GoWriter writer = context.getWriter();
+        SmithyPattern pattern = operation.expectTrait(EndpointTrait.class).getHostPrefix();
+
+        // If the pattern is just a string without any labels, then we simply use string
+        // assignment to avoid unnecessary imports / work.
+        if (pattern.getLabels().isEmpty()) {
+            writer.write("request.HostPrefix = $S", pattern.toString());
+            return;
+        }
+
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        StructureShape input = ProtocolUtils.expectInput(context.getModel(), operation);
+
+        // If the pattern has labels, we need to build up the host prefix using a string builder.
+        writer.addUseImports(SmithyGoDependency.STRINGS);
+        writer.write("var prefix strings.Builder");
+        for (Segment segment : pattern.getSegments()) {
+            if (!segment.isLabel()) {
+                writer.write("prefix.WriteString($S)", segment.toString());
+            } else {
+                MemberShape member = input.getMember(segment.getContent()).get();
+                String memberReference = "input." + symbolProvider.toMemberName(member);
+
+                // Theoretically this should never be nil by this point unless validation has been disabled.
+                writer.openBlock("if $L != nil {", "}", memberReference, () -> {
+                    writer.write("prefix.WriteString(*$L)", memberReference);
+                });
```

Comment:
This probably should return an error if the member is nil, its possible for this to fail silently. This is one of those checks that probably should be required even if validation is disabled, like URL path label replaces. e.g. GetObject with validation disabled and no-bucket, will result in an operation that does a ListObjects instead.

Also this should validate that the member being used is a valid host label, (V1's implementation for reference)

https://github.com/aws/aws-sdk-go/blob/master/private/protocol/host.go#L51-L68